Getting rid of `Makefile.tools`

shindere opened the pull request:
So far, this file has been used to let the testsuite and the manual access
the in-tree compilers (well, it's only the build of the supporting
libraries and tools of the testsuite which were relying on it, not
ocamltest itself).

This PR removes this file and its dedicated variables, so that the
manual and the testsuite rely on the same variables than the rest of
the build system to access the compilers.

As a small side-benefit, the manual now benefits from the
`Makefile.best_binaries` enhancements.

Cc @Octachron for the manual-related part, also @dra27 may want to
have a look, especially because I have no idea whether I ddin't
break things when flexdll is bootstrapped -- hopefully not!

that being said I ran the branch through precheck, see build #557, and
things seem to (still) work nicely.

shindere committed:
Remove the unused CTOPDIR build variable

shindere committed:
Makefile.tools & co: get rid of the TOPDIR variable

It's a synonymous for ROOTDIR so rather use this later one as in the
rest of the build system.

shindere committed:
Makefile.tools: remove the unused UNIXLIBVAR variable

shindere committed:
Makefile.tools: remove the unused DUMPOBJ and OBJINFO variables

shindere committed:
Makefile.tools: remove the unused OCAMLMKLIB variable

shindere committed:
Makefile.tools: remove the unused DIFF variable

shindere committed:
Makefile.tools: get rid of the unused OCOPTFLAGS variable

shindere committed:
Makefile.tools: remove the unused OCAML variable

shindere committed:
testsuite/lib/Makefile: stop using Makefile.tools

shindere committed:
testsuite/tools/Makefile: stop using Makefile.tools

shindere committed:
testsuite/Makefile: stop using Makefile.tools

This requires that Makefile.config.in defines the SORT, CYGPATH and
SET_LD_LIBRARY_PATH variables also on Unix, not only on Windows as before.

It is also necessary to set OCAMLRUNPARAM to v=0 when either the
instrumented or the debug runtime are used, so that their output
does not interfer with test output. Before this commit this was
doner in Makefile.tools but now it's done in testsuite/Makefile.

OCAMLRUNPARAM fix

shindere committed:
Stop using Makefile.tools in the makefiles used to build the manual

shindere committed:
Remove Makefile.tools, now unused.

dra27 commented:
There are various new references to `Makefile.local` in the manual's build system - is that an accidental change, or is there a new file which you need to push?

shindere commented:
David Allsopp (2021/04/12 12:53 -0700):
> There are various new references to `Makefile.local` in the manual's
> build system - is that an accidental change, or is there a new file
> which you need to push?

Oops that was a mistake, thanks!

It should have been `Makefile.common`, of course.

It has now been fixed.

dra27 commented:
Ah, I didn't think of that! There still seems to be something going on with the no-naked-pointers check (there was before as well). It's just doing:
```
./configure --disable-naked-pointers --disable-stdlib-manpages --disable-dependency-generation --enable-ocamltest
make -j world.opt
make -C testsuite USE_RUNTIME=d all
```
but is getting a large amount of output (the logs are really difficult to inspect online because of their size - I expect it'll be easier to poke locally).

shindere commented:
Many thansk for having provided this feedback, @dra27!

I'm not so surprised that this goes wrong becuase `USE_RUNTIME` is part
of the varialbes that was defined in `Makefile.tools`.

I am going to investigate this right now and will report back here as
soon as possible.

Thanks again for having given the steps to reproduce!

I was also wondering how easy it would be to run all the GitHub actions
locally? Is that even possible? I didn't investigate so far.

dra27 commented:
> I was also wondering how easy it would be to run all the GitHub actions locally? Is that even possible? I didn't investigate so far.

I've not actually tested this, but the intention is that `.github/workflows/build.yml` calls into `tools/ci/actions/runner.sh` passing any GitHub magic needed as parameters, so you should be able to execute `runner.sh` locally without having to worry about any magic environment variables and so forth.

shindere commented:
David Allsopp (2021/04/13 01:30 -0700):
> Ah, I didn't think of that! There still seems to be something going on with the no-naked-pointers check (there was before as well). It's just doing:
> ```
> ./configure --disable-naked-pointers --disable-stdlib-manpages --disable-dependency-generation --enable-ocamltest
> make -j world.opt
> make -C testsuite USE_RUNTIME=d all
> ```
> but is getting a large amount of output (the logs are really difficult
> to inspect online because of their size - I expect it'll be easier to
> poke locally).

I think this should now be fixed.

the problem was that, when using either the instrumented or debug
runtime, `OCAMLRUNPARAM` was set to `v+0` by default in `Makefile.tools`
and I didn't transport that to `testsuite/Makefile`, which I have now
done.

Actually, one other possibility would be to set OCAMLRUNPARAM correctly
directly in ocamltest but that can perhaps be done in a separate PR
since the current modification can be considered as more faithful, in
spirit, to what is going on here (just getting rid of a Makefile).

Also, I think dealing with this in the testsuite's Makefile is an
improvement over dealing with it in a generic Makefile as was done
before.

dra27 commented:
`make -C manual/tools` is failing because there's no active definition of `OCAMLLEX`. It used to get that from `Makefile.tools` but `Makefile.common` uses it without defining it. It's scattered all over the build system. Up to you whether to factorise it fully or just put `OCAMLLEX ?= $(BEST_OCAMLLEX)` in `src/tools/Makefile`

shindere commented:
David Allsopp (2021/04/13 03:13 -0700):
> `make -C manual/tools` is failing because there's no active definition
> of `OCAMLLEX`. It used to get that from `Makefile.tools` but
> `Makefile.common` uses it without defining it. It's scattered all over
> the build system. Up to you whether to factorise it fully or just put
> `OCAMLLEX ?= $(BEST_OCAMLLEX)` in `src/tools/Makefile`

Thanks! I went for your second approach because the "right" value to use
for OCAMLLEX is not always the same in all directories.

One other way could be to use this default value in Makefile.common, so that
it's sure by construction that it is always defined, and then override
it in those directories where it's different. But that does not quite
work because then, I think, it's no more possible to also override it
when calling make, which is something important to keep, I think.

dra27 commented:
> One other way could be to use this default value in Makefile.common, so that it's sure by construction that it is always defined, and then override it in those directories where it's different. But that does not quite work because then, I think, it's no more possible to also override it when calling make, which is something important to keep, I think.

This could be kept with this in `Makefile.common` (I haven't tested it on macOS, but I checked in the GNU make 3.80 manual for the functions!):

```make
%.ml: %.mll
	$(if $(filter undefined,$(origin OCAMLLEX)),$(BOOT_OCAMLLEX),$(OCAMLLEX)) $(OCAMLLEXFLAGS) $<
```

That will cause `Makefile.common` to use `BOOT_OCAMLLEX` by default while also allowing individual `Makefile`s to have `OCAMLLEX ?= ...` (and it doesn't trigger the `-warn-undefined-variables` test)

shindere commented:
David Allsopp (2021/04/13 04:21 -0700):
> > One other way could be to use this default value in Makefile.common, so that it's sure by construction that it is always defined, and then override it in those directories where it's different. But that does not quite work because then, I think, it's no more possible to also override it when calling make, which is something important to keep, I think.
>
> This could be kept with this in `Makefile.common` (I haven't tested it on macOS, but I checked in the GNU make 3.80 manual for the functions!):
>
> ```make
> %.ml: %.mll
> 	$(if $(filter undefined,$(origin OCAMLLEX)),$(BOOT_OCAMLLEX),$(OCAMLLEX)) $(OCAMLLEXFLAGS) $<
> ```
>
> That will cause `Makefile.common` to use `BOOT_OCAMLLEX` by default
> while also allowing individual `Makefile`s to have `OCAMLLEX ?= ...`
> (and it doesn't trigger the `-warn-undefined-variables` test)

Can you have a look to the last commit I just pushed?

I think it fixes things in a nice way. What do you think?

dra27 commented:
I'm not convinced, to be honest - the comment needs to be where `OCAMLLEX` is overridden (it almost looks like an error that `OCAMLLEX` is defined at the top of `Makefile` and `lex/Makefile` yet all the other definitions (`CAMLDEP`, etc.) are further down the file. If someone were to move it, I think the change in semantics would be quite subtle to debug as well.

The `$(origin` version may use more make trickery, but it at least has the benefit that it never gets changed and the other files can define `OCAMLLEX` without constraints on where.

shindere commented:
Well, I really prefer this version and think we shall move in this
direction further, until of course we move to a main Makefile that
includes the sub-makefiles rather than calling them.

I find the version with origin more difficult to read and I don't really
like the complexity added to the recipe in the `%.ml: %.mll` rule. To me
this rule should be kept simple and straightforward because the hcoice
of the right lexer to use depends on the context and I would like this
rule to know nothing aobut the context.

In case you really don't like this last commit,  I could also submit the
PR without it, although I personnally see it as an improvement.

dra27 commented:
No, that's fine - but can the comment about placement therefore be in `tools/Makefile` and `Makefile`?

shindere commented:
David Allsopp (2021/04/13 08:47 -0700):
> No, that's fine - but can the comment about placement therefore be in
> `tools/Makefile` and `Makefile`?

Sure, will do -- tomorrow. And in `lex/Makefile`, too, I guess. That's a
good idea to put the comment there. If you don't mind I propose to leave
it in `Makefile.common` rather than removing it from this place.

shindere commented:
David Allsopp (2021/04/13 08:47 -0700):
> No, that's fine - but can the comment about placement therefore be in
> `tools/Makefile` and `Makefile`?

Done! I updated thelatest commit to include the comment where
appropriate.

Many thanks, David for being so responsive and constructive!

dra27 commented on Makefile.common:
```diff
@@ -127,10 +127,12 @@ endef # PROGRAM_SYNONYM
-# freshly compiled lexer generator.
+
+# Default value for OCAMLLEX
+# In those directories where this needs to be override, the overriding
```
```suggestion
# In those directories where this needs to be overridden, the overriding
```
(typo)

dra27 commented on lex/Makefile:
```diff
@@ -16,7 +16,11 @@
 # The lexer generator
 
 ROOTDIR = ..
+# NOTE: it is important that OCAMLLEX is defined *before* Makefile.common
+# gets included, so that its definition here takes precedence
+# over the one there.
 
```
```suggestion

# NOTE: it is important that OCAMLLEX is defined *before* Makefile.common
# gets included, so that its definition here takes precedence
# over the one there.
```
(whitespace adjustment)

dra27 commented on manual/tools/Makefile:
```diff
@@ -1,28 +1,27 @@
-	$(OCAMLLEX) $<
-
-
+	$(OCAMLC) $(COMPFLAAGS) -c $<
```
```suggestion
	$(OCAMLC) $(COMPFLAGS) -c $<
```

dra27 commented on testsuite/Makefile:
```diff
@@ -93,6 +93,17 @@ OCAMLTESTFLAGS := \
 
+ifneq ($(USE_RUNTIME),)
+# Check USE_RUNTIME value
+ifeq ($(findstring $(USE_RUNTIME),d i),)
+$(error If set, USE_RUNTIME must be equal to "d" (debug runtime) \
+        or "i" (instrumented runtime))
```
```suggestion
ifeq ($(wildcard $(ROOTDIR)/runtime/ocamlrun$(USE_RUNTIME),)
$(error USE_RUNTIME=$(USE_RUNTIME) but \
        $(ROOTDIR)/runtime/ocamlrun$(USE_RUNTIME) not found.)
```
rather than hard-coding the list? It also guards against trying to run with `USE_RUNTIME=d` and `--disable-debug-runtime` (I realise it's a change from what was there before)

dra27 commented on testsuite/Makefile:
```diff
@@ -93,6 +93,17 @@ OCAMLTESTFLAGS := \
+endif
+# When using the debug or instrumented runtime,
+# set the runtime's verbosity to 0 by default
+export OCAMLRUNPARAM?=v=0
```
Conversely:
```suggestion
ifneq ($(findstring $(USE_RUNTIME),d i),)
export OCAMLRUNPARAM?=v=0
endif
```

dra27 commented on manual/tools/Makefile:
```diff
@@ -1,28 +1,27 @@
-COMPFLAGS=-I $(OTOPDIR)/otherlibs/str -I $(OTOPDIR)/otherlibs/unix
-include $(TOPDIR)/Makefile.tools
+ROOTDIR = ../..
+COMPFLAGS=-I $(OROOTDIR)/otherlibs/str -I $(OROOTDIR)/otherlibs/unix
```
```suggestion
COMPFLAGS=-I $(ROOTDIR)/otherlibs/str -I $(ROOTDIR)/otherlibs/unix
```

dra27 commented on testsuite/Makefile:
```diff
@@ -93,6 +93,17 @@ OCAMLTESTFLAGS := \
   $(OCAMLTEST_PROMOTE_FLAG) \
   $(OCAMLTEST_KEEP_TEST_DIR_ON_SUCCESS_FLAG)
 
+ifneq ($(USE_RUNTIME),)
```
From `Makefile.tools` too:
```suggestion
# Make sure USE_RUNTIME is defined
USE_RUNTIME ?=

ifneq ($(USE_RUNTIME),)
```

dra27 reviewed (commented):
A few typos, and - while the code's moving from `Makefile.tools` - a possible improvement to the detection for `USE_RUNTIME`.

It's very nice to be seeing the back of this largely duplicating `Makefile`!

dra27 commented:
(I've put the `no-change-entry-needed` label on, but that doesn't preclude you adding a `Changes` entry if you wish, of course!)

shindere commented:
Thanks David!

I think I'm fine with no change entry, atcually.

I'm gonna try to integrate your comemnts. I think I like all of them so
it's not about not wanting to integrate them, it's just that, when
reading the comment as an e-mail text, I am having a hard time to
identify to which file a comment corresponds.

One other thing I have a hard time figuring out is what a suggestion
actually means. I noeed to go through my own writing and then through
the suggesiton and it's sometimes (very) painful to find out what the
difference actually is. So if the suggestions could describe the changes
rather than just write what they should be, it would I think make things
a bit easier for me.

But I'll try and let you know what I cannot identify.

shindere commented:
David Allsopp (2021/04/13 11:48 -0700):
> @dra27 commented on this pull request.
>
> A few typos, and - while the code's moving from `Makefile.tools` - a
> possible improvement to the detection for `USE_RUNTIME`.

Nice! Thanks!

> It's very nice to be seeing the back of this largely duplicating
> `Makefile`!

thank you!

> > @@ -127,10 +127,12 @@ endef # PROGRAM_SYNONYM
>  # Lexer generation
>
>  BOOT_OCAMLLEX ?= $(CAMLRUN) $(ROOTDIR)/boot/ocamllex
> -# The OCAMLLEX command used in the recipe below is defined in the individual
> -# makefiles, because its default value is directory-specific.
> -# By default, some directories use BOOT_OCAMLLEX while others use the
> -# freshly compiled lexer generator.
> +
> +# Default value for OCAMLLEX
> +# In those directories where this needs to be override, the overriding
>
> ```suggestion
> # In those directories where this needs to be overridden, the overriding
> ```
> (typo)

Fixed, thanks. that's a good example of difficult situation: it took me
a while to figure out what was wrong.

> > +# NOTE: it is important that OCAMLLEX is defined *before* Makefile.common
> +# gets included, so that its definition here takes precedence
> +# over the one there.
>
>
> ```suggestion
>
> # NOTE: it is important that OCAMLLEX is defined *before* Makefile.common
> # gets included, so that its definition here takes precedence
> # over the one there.
> ```
> (whitespace adjustment)

Here I couldn't perceive what had been suggested. For what I could
perceive, the text and its suggested replacement were identical, e.g.
the way the text was split into lines felt the same.

>  %.cmi: %.mli
> -	$(OCAMLC) -c $<
> -
> -%.ml: %.mll
> -	$(OCAMLLEX) $<
> -
> -
> +	$(OCAMLC) $(COMPFLAAGS) -c $<
>
> ```suggestion
> 	$(OCAMLC) $(COMPFLAGS) -c $<
> ```

Same here, what exactly is it that you are suggesting? In which file?

> > +ifeq ($(findstring $(USE_RUNTIME),d i),)
> +$(error If set, USE_RUNTIME must be equal to "d" (debug runtime) \
> +        or "i" (instrumented runtime))
>
> ```suggestion
> ifeq ($(wildcard $(ROOTDIR)/runtime/ocamlrun$(USE_RUNTIME),)
> $(error USE_RUNTIME=$(USE_RUNTIME) but \
>         $(ROOTDIR)/runtime/ocamlrun$(USE_RUNTIME) not found.)
> ```
> rather than hard-coding the list? It also guards against trying to run with `USE_RUNTIME=d` and `--disable-debug-runtime` (I realise it's a change from what was there before)
>
> > @@ -93,6 +93,17 @@ OCAMLTESTFLAGS := \
>    $(OCAMLTEST_PROMOTE_FLAG) \
>    $(OCAMLTEST_KEEP_TEST_DIR_ON_SUCCESS_FLAG)
>
> +ifneq ($(USE_RUNTIME),)
> +# Check USE_RUNTIME value
> +ifeq ($(findstring $(USE_RUNTIME),d i),)
> +$(error If set, USE_RUNTIME must be equal to "d" (debug runtime) \
> +        or "i" (instrumented runtime))
> +endif
> +# When using the debug or instrumented runtime,
> +# set the runtime's verbosity to 0 by default
> +export OCAMLRUNPARAM?=v=0
>
> Conversely:
> ```suggestion
> ifneq ($(findstring $(USE_RUNTIME),d i),)
> export OCAMLRUNPARAM?=v=0
> endif
> ```

But the list has still to be hard-coded, hasn't it?

> > @@ -1,28 +1,27 @@
> -TOPDIR=../..
> -COMPFLAGS=-I $(OTOPDIR)/otherlibs/str -I $(OTOPDIR)/otherlibs/unix
> -include $(TOPDIR)/Makefile.tools
> +ROOTDIR = ../..
> +COMPFLAGS=-I $(OROOTDIR)/otherlibs/str -I $(OROOTDIR)/otherlibs/unix
>
> ```suggestion
> COMPFLAGS=-I $(ROOTDIR)/otherlibs/str -I $(ROOTDIR)/otherlibs/unix
> ```

Fixed, thanks.

dra27 commented:
Sorry, I had no idea the email version was so inaccessible! I'll aim not to use those and push a reference to a commit or post the diff directly in future! For these:

`lex/Makefile` - my suggestion was to remove the blank line between the comment and `OCAMLLEX ?=` and insert a blank line above the comment (or "move the blank line", if you like!!).

`manual/tools/Makefile` - you've ended up with `$(COMPFLAAGS)` instead of `$(COMPFLAGS)` (stray extra `A`)

For the `USE_RUNTIME` validation, I was suggesting switching the validation of the parameter from a hard-coded list to `$(wildcard` detection. It's pedantically resilient against future runtime variants but more importantly it also adds an extra check to ensure that the debug or instrumented runtimes have actually been built. Hard-coding the list of variants for which the `v=0` setting is required is acceptable - in this case, we explicitly know that those two runtimes set a different default; any future runtime variant would need adding, we can't guess that in advance.

shindere committed:
Build system: provide a default value for OCAMLLEX

Each directory can override this default value.

shindere commented:
David Allsopp (2021/04/13 12:44 -0700):
> Sorry, I had no idea the email version was so inaccessible!

No problem at all, you couldn't dream it! :)

It actually took me a while to realise that others got a much better
result than I had.

> I'll aim not to use those and push a reference to a commit or post the diff directly in future! For these:

Well, the way you just did is convenient for me, but perhaps it's too
cumbersome for you...

Diffs are actually not particularly easy to read in braille either. Even
whenthe old and new lines are consecutive, it's sometimes not obvious
where the change is. And, if the lines are not consecutive in the diff,
if there are, say, 3 lines removed and 3 others added, then it becomes
very very tedious to sort out what's going on.

> `lex/Makefile` - my suggestion was to remove the blank line between
> the comment and `OCAMLLEX ?=` and insert a blank line above the
> comment (or "move the blank line", if you like!!).

Done, thanks. I also added one extra blank line between the definition
of OCAMLLEX and the include directive.

> `manual/tools/Makefile` - you've ended up with `$(COMPFLAAGS)` instead
> of `$(COMPFLAGS)` (stray extra `A`)

Fixed! Thanks!

> For the `USE_RUNTIME` validation, I was suggesting switching the
> validation of the parameter from a hard-coded list to `$(wildcard`
> detection. It's pedantically resilient against future runtime variants
> but more importantly it also adds an extra check to ensure that the
> debug or instrumented runtimes have actually been built.

I see. Since it's a slightly deeper change I will do that one tomorrow.
I understand the suggestion anyway and like it. thanks a lot.

> Hard-coding
> the list of variants for which the `v=0` setting is required is
> acceptable - in this case, we explicitly know that those two runtimes
> set a different default; any future runtime variant would need adding,
> we can't guess that in advance.

OK I understand, you suggest to dissociate the two things and it makes
sense to me, of course. That, too, I'll do tomorrow.

If you see other thigns you'd like me to fix / improve, odn't hesitate
to let me know so that I can integrate those in my next iteration.

And thanks again for being so supportive!

shindere commented:
David Allsopp (2021/04/13 12:44 -0700):
> For the `USE_RUNTIME` validation, I was suggesting switching the
> validation of the parameter from a hard-coded list to `$(wildcard`
> detection. It's pedantically resilient against future runtime variants
> but more importantly it also adds an extra check to ensure that the
> debug or instrumented runtimes have actually been built.

Well, giving this a second thought, it occurred to me that this may not
be such a good idea, actually. Two reasons for that:

 1. The logic you describe is already hard-wired in ocamltest, see the runtime_variant function in `ocaml_files.ml` so your proposal would mean hardwiring the same logic in two places (after all, the only thing this Makefile does is calling ocamltest so the fact that the logic is in the Makefile does not seem to add any significant benefit.

 2. It happened to me that I had the need to use other runtimes than those hard-coded in ocamltest and I think this will become a real use-case, e.g. in the context of cross-compilation which is what triggered this PR in the first place. This consideration makes me even more reluctent to duplicate a logic when I can already see its limitations.



> Hard-coding the list of variants for which the `v=0` setting is required is
> acceptable - in this case, we explicitly know that those two runtimes
> set a different default; any future runtime variant would need adding,
> we can't guess that in advance.

Indeed. But then, given the above, I am wondering where this should be
hard-coded. And I am suspecting it would make more sense to add this to
ocamltest itself than to the makefile, in particular because it would
guarantee a consistency of behaviours between running the tests through
make and calling ocamltest directly. I have to say I did think about
this point when working on the PR but that I initially renounced to
include the modification, partly because I wanted to minimize the diff.
However, now that we are there, if you agree that this approach would be
more sensible, I wouldn't mind including a modification so that it's
ocamltest which exports `OCAMLRUNPARAM=v=0` when either the instrumented
or the debug runtime is used, unless of course the varialbe already
exists in the environment, in which case it would preserve it, I think
that's what the current implementation does.

dra27 commented:
There are two parts to the validation: I agree with you on the second that the setting of `OCAMLRUNPARAM=v=0` would be much better delegated to `ocamltest` but on the first part, the change I'm proposing is not presently in ocamltest. `Ocaml_files.runtime_variant` checks whether `USE_RUNTIME` is `d` or `i` (and silently ignores it otherwise). I'm proposing here a check somewhere that the runtime has actually been built. At present, `./configure --disable-debug-runtime --disable-native-compiler && make -j && make -C testsuite USE_RUNTIME=d all` results in a sea of output and eventually 1113 tests failed where I'm suggesting a single error message that the runtime isn't built followed by no tests run.

All that said, I wonder if it would be better to get this PR in as it is - I originally suggested those two changes because they looked new (i.e. I didn't realise it was code _moving_ from `Makefile.tools` initially). The only change needed if we do that is to insert these two lines from `Makefile.tools` above line 96 of `testsuite/Makefile` to pedantically satisfy the `-warn-unused-variables` aim:

```
# Make sure USE_RUNTIME is defined
USE_RUNTIME ?=
```

For a future change, in general, I dislike environment variables controlling programs (because the syntax for doing that on Windows is very cumbersome) - I think it would be better for ocamltest to have a command line parameter for selecting the runtime (e.g. `-use-runtime`) and that it then does _all_ of the validation and setting of `OCAMLRUNPARAM` and then have `testsuite/Makefile` convert `$(USE_RUNTIME)` into `-use-runtime=$(USE_RUNTIME)` (i.e. `USE_RUNTIME` is a parameter to `testsuite/Makefile`, rather than `ocamltest`). What do you think?

shindere commented:
David Allsopp (2021/04/14 01:37 -0700):
> There are two parts to the validation: I agree with you on the second
> that the setting of `OCAMLRUNPARAM=v=0` would be much better delegated
> to `ocamltest`

Nice!

> but on the first part, the change I'm proposing is not
> presently in ocamltest. `Ocaml_files.runtime_variant` checks whether
> `USE_RUNTIME` is `d` or `i` (and silently ignores it otherwise).

Yes yes, I was aware of that and would have proposed to update this to
not silently ignore an unknown setting.

> I'm proposing here a check somewhere that the runtime has actually been
> built. At present, `./configure --disable-debug-runtime
> --disable-native-compiler && make -j && make -C testsuite
> USE_RUNTIME=d all` results in a sea of output and eventually 1113
> tests failed where I'm suggesting a single error message that the
> runtime isn't built followed by no tests run.

Yes, I agree on this. That's exsactly why I wrote that the change would
not bring a "significant" benefit, but I agree that it would be better
to check USE_RUNTIME once and for all and not let ocamltest check it
each time it's run.

> All that said, I wonder if it would be better to get this PR in
> as it is - I originally suggested those two changes because they
> looked new (i.e. I didn't realise it was code _moving_ from
> `Makefile.tools` initially). The only change needed if we do that
>is to insert these two lines from `Makefile.tools` above line 96
> of `testsuite/Makefile` to pedantically satisfy the
> `-warn-unused-variables` aim:
>
> ```
> # Make sure USE_RUNTIME is defined
> USE_RUNTIME ?=
> ```

That has now been done. Well spotted! thanks!

> For a future change, in general, I dislike environment variables
> controlling programs (because the syntax for doing that on Windows is
> very cumbersome) - I think it would be better for ocamltest to have a
> command line parameter for selecting the runtime (e.g. `-use-runtime`)
> and that it then does _all_ of the validation and setting of
> `OCAMLRUNPARAM` and then have `testsuite/Makefile` convert
> `$(USE_RUNTIME)` into `-use-runtime=$(USE_RUNTIME)` (i.e.
> `USE_RUNTIME` is a parameter to `testsuite/Makefile`, rather than
> `ocamltest`). What do you think?

I like the idea in general, yes!

The only thing I would do a bit differently is that I would perhaps
suggest to provide both the environment variable andthe command-line
option, the second one taking precedence over the first one.

We could also have another ocamltest option to specify a runtime in a
more concrete way, e.g. a path to an executable or somethign like that,
to have more flexibility.

dra27 reviewed (approved):
Good to go with CI - the i386-static failure looks to be that fixed in #10334

Octachron commented:
There seems to be a build failure on the manual due to an undefined CAMLRUN in manual/src/cmds/Makefile.

shindere commented:
Florian Angeletti (2021/04/14 02:59 -0700):
> There seems to be a build failure on the manual due to an undefined
> CAMLRUN in manual/src/cmds/Makefile.

Thanks, Florian! Can you provide steps to reproduce this, please?

Octachron commented:
`./configure && make -j && make -C manual` is sufficient to reproduce the issue for me.

dra27 commented:
I'm sorry, I was sure I'd run this manually (excuse pun), but clearly not! It looks like a lot of `s/OCAMLRUN/CAMLRUN` is sufficient:

```diff
diff --git a/manual/src/cmds/Makefile b/manual/src/cmds/Makefile
index b71c1911d..3b6964177 100644
--- a/manual/src/cmds/Makefile
+++ b/manual/src/cmds/Makefile
@@ -5,10 +5,10 @@ LD_PATH = "$(ROOTDIR)/otherlibs/str:$(ROOTDIR)/otherlibs/unix"

 TOOLS = ../../tools
 CAMLLATEX = $(SET_LD_PATH) \
-  $(OCAMLRUN) $(ROOTDIR)/tools/caml-tex \
+  $(CAMLRUN) $(ROOTDIR)/tools/caml-tex \
   -repo-root $(ROOTDIR) -n 80 -v false
-TEXQUOTE = $(OCAMLRUN) $(TOOLS)/texquote2
-TRANSF = $(SET_LD_PATH) $(OCAMLRUN) $(TOOLS)/transf
+TEXQUOTE = $(CAMLRUN) $(TOOLS)/texquote2
+TRANSF = $(SET_LD_PATH) $(CAMLRUN) $(TOOLS)/transf

 FILES = comp.tex top.tex runtime.tex native.tex lexyacc.tex intf-c.tex \
   ocamldep.tex profil.tex debugger.tex ocamldoc.tex \
diff --git a/manual/src/refman/Makefile b/manual/src/refman/Makefile
index 7b03ad0c9..e8dbfb1ab 100644
--- a/manual/src/refman/Makefile
+++ b/manual/src/refman/Makefile
@@ -5,10 +5,10 @@ LD_PATH = "$(ROOTDIR)/otherlibs/str:$(ROOTDIR)/otherlibs/unix"

 TOOLS = ../../tools
 CAMLLATEX = $(SET_LD_PATH) \
-  $(OCAMLRUN) $(ROOTDIR)/tools/caml-tex \
+  $(CAMLRUN) $(ROOTDIR)/tools/caml-tex \
   -repo-root $(ROOTDIR) -n 80 -v false
-TEXQUOTE = $(OCAMLRUN) $(TOOLS)/texquote2
-TRANSF = $(SET_LD_PATH) $(OCAMLRUN) $(TOOLS)/transf
+TEXQUOTE = $(CAMLRUN) $(TOOLS)/texquote2
+TRANSF = $(SET_LD_PATH) $(CAMLRUN) $(TOOLS)/transf

 EXTENSION_FILES = letrecvalues.tex recursivemodules.tex locallyabstract.tex \
   firstclassmodules.tex moduletypeof.tex signaturesubstitution.tex \
diff --git a/manual/src/tutorials/Makefile b/manual/src/tutorials/Makefile
index 1061b9e15..5ce1fefd2 100644
--- a/manual/src/tutorials/Makefile
+++ b/manual/src/tutorials/Makefile
@@ -5,10 +5,10 @@ LD_PATH = "$(ROOTDIR)/otherlibs/str:$(ROOTDIR)/otherlibs/unix"

 TOOLS = ../../tools
 CAMLLATEX = $(SET_LD_PATH) \
-  $(OCAMLRUN) $(ROOTDIR)/tools/caml-tex \
+  $(CAMLRUN) $(ROOTDIR)/tools/caml-tex \
   -repo-root $(ROOTDIR) -n 80 -v false
-TEXQUOTE = $(OCAMLRUN) $(TOOLS)/texquote2
-TRANSF = $(SET_LD_PATH) $(OCAMLRUN) $(TOOLS)/transf
+TEXQUOTE = $(CAMLRUN) $(TOOLS)/texquote2
+TRANSF = $(SET_LD_PATH) $(CAMLRUN) $(TOOLS)/transf


 FILES = coreexamples.tex lablexamples.tex objectexamples.tex \
diff --git a/manual/tests/Makefile b/manual/tests/Makefile
index ec3cd499c..435cd3fc7 100644
--- a/manual/tests/Makefile
+++ b/manual/tests/Makefile
@@ -23,7 +23,7 @@ cross-reference-checker: cross_reference_checker.ml
 .PHONY: check-cross-references
 check-cross-references: cross-reference-checker
        $(SET_LD_PATH) \
-         $(OCAMLRUN) ./cross-reference-checker \
+         $(CAMLRUN) ./cross-reference-checker \
          -auxfile $(MANUAL)/texstuff/manual.aux \
          $(ROOTDIR)/utils/warnings.ml \
          $(ROOTDIR)/driver/main_args.ml \
```

shindere commented:
Thanks a lot for your help and support @dra27 and @Octachron, and very
sorry I didn't do things correctly right from the beginning.

David Allsopp (2021/04/14 03:35 -0700):
> I'm sorry, I was sure I'd run this manually (excuse pun),

:-)

> but clearly not! It looks like a lot of `s/OCAMLRUN/CAMLRUN` is sufficient:
[...]

It works indeed!

I would have prefered to do the renaming the other way around, but I
think that would have been more difficult so I'll keep that battle for
another day.

PR updated, building the manual now works for me whereas I could
reproduce the issue before.

dra27 commented:
> I would have prefered to do the renaming the other way around, but I think that would have been more difficult so I'll keep that battle for another day.

Me too! I'll be right beside you for that battle :o)

dra27 commented:
> > I would have prefered to do the renaming the other way around, but I think that would have been more difficult so I'll keep that battle for another day.
> 
> Me too! I'll be right beside you for that battle :o)

Maybe that change could be done for OCaml's 25th birthday next month!

shindere commented:
David Allsopp (2021/04/14 07:25 -0700):
> > I would have prefered to do the renaming the other way around, but I think that would have been more difficult so I'll keep that battle for another day.
>
> Me too! I'll be right beside you for that battle :o)

Thanks! I expect to be able to work on that soon, so that allthe build
variables get consistent names and get used consistently in the whole
build system.

Thanks for your support! Very warmly appreciated!

shindere commented:
David Allsopp (2021/04/14 07:26 -0700):
> > > I would have prefered to do the renaming the other way around, but I think that would have been more difficult so I'll keep that battle for another day.
> >
> > Me too! I'll be right beside you for that battle :o)
>
> Maybe that change could be done for OCaml's 25th birthday next month!

Why not!

Octachron reviewed (approved):
The simplifications on the manual side looks good to me (now that the build is fixed).

shindere commented:
David Allsopp (2021/04/14 08:10 -0700):
> Merged #10335 into trunk.

Hurra! And thanks!!!
